### PR TITLE
DOCSP-9587 builders updates

### PR DESCRIPTION
**Pull request by kyuan-mongodb**
### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-9587

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodbcom-staging.corp.mongodb.com/652a724/java/docsworker-xlarge/DOCSP-9587-BuildersUpdates/fundamentals/builders/updates/

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- [x] Are the staging links in the PR description updated?

### CODE REVIEW

**Review by terakilobyte — CHANGES_REQUESTED** [review]
This looks really good. I just have a few questions. One is in a review comment, and the other is:

When I read through the examples and how the document changes, the wording is
`The example above updates the document to the following state:`. Could we clarify this by adding something like `... updates the original document...` or making them build upon on another? That way the user doesn't have to keep the original document state in mind, or scroll up if they are incredulous.

File: source/fundamentals/builders/updates.txt
Comment (kyuan-mongodb) [reply source/fundamentals/builders/updates.txt:114]: I actually tried it out in my playground, and if an update doesn't occur, nothing happens

Comment (kyuan-mongodb) [reply source/fundamentals/builders/updates.txt:114]: I added a note about this on the page.  I also want to keep the examples to the original document in the case a user on the page is looking for a particular example.

**Review by terakilobyte — APPROVED** [review]
LGTM

File: source/fundamentals/builders/updates.txt
Comment (jyemin) [inline source/fundamentals/builders/updates.txt:280]: To update it to a timestamp, you have to use `currentTimestamp` I think

File: source/fundamentals/builders/updates.txt
Comment (jyemin) [inline source/fundamentals/builders/updates.txt:436]: Does it remove all or just the first?  As is, the description of this method is the same as the description for pullAll. (I think the Javadoc has the same confusion, but it would be good to clarify here and perhaps also fix the Javadoc)

**Review by jyemin — CHANGES_REQUESTED** [review]

File: source/fundamentals/builders/updates.txt
Comment (kyuan-mongodb) [reply source/fundamentals/builders/updates.txt:50]: That's what printed out when I performed a System.out.print

Comment (jyemin) [reply source/fundamentals/builders/updates.txt:50]: Right.  But that uses the `toString()` method, which doesn't produce JSON.  If you use the `toJson()` method instead, you'll get something like: `{"$date": "2021-03-18T19:18:17.265Z"}` for any date value.

Comment (kyuan-mongodb) [reply source/fundamentals/builders/updates.txt:50]: Correct, I just checked and removed my comment since I saw what you meant.

File: source/fundamentals/builders/updates.txt
Comment (kyuan-mongodb) [reply source/fundamentals/builders/updates.txt:280]: https://docs.mongodb.com/manual/reference/operator/update/currentDate/

Comment (jyemin) [reply source/fundamentals/builders/updates.txt:280]: Right, but the currentDate builder doesn't work that way: it only generates dates.  You have to use currentTimestamp to generate a timestamp.  Maybe not worth mentioning that though since you rarely need a timestamp.

File: source/fundamentals/builders/updates.txt
Comment (kyuan-mongodb) [reply source/fundamentals/builders/updates.txt:436]: So the difference is that pull only removes one value, but pullAll can do multiple. I didn't properly convey it in this PR, so I'll be updating it

**Review by jyemin — COMMENTED** [review]
Waiting for request for second review

**Comment by kyuan-mongodb** [comment]
@jyemin This is ready for a re-review

File: source/fundamentals/builders/updates.txt
Comment (jyemin) [reply source/fundamentals/builders/updates.txt]: Looks like you replaced it with the shell syntax.  I think it would be better to use extended JSON syntax instead.

File: source/fundamentals/builders/updates.txt
Comment (jyemin) [reply source/fundamentals/builders/updates.txt:286]: What did you end up doing with this one?

File: source/fundamentals/builders/updates.txt
Comment (jyemin) [reply source/fundamentals/builders/updates.txt:442]: What did you end up doing with this one?

File: source/includes/fundamentals/code-snippets/builders/Updates.java
Comment (jyemin) [inline source/includes/fundamentals/code-snippets/builders/Updates.java:43]: Did you intend to leave the println statements in?

Comment (kyuan-mongodb) [reply source/includes/fundamentals/code-snippets/builders/Updates.java:43]: Yes, I forgot to include them before and it took me a while to figure out the order of the updates being performed.

File: source/fundamentals/builders/updates.txt
Comment (kyuan-mongodb) [reply source/fundamentals/builders/updates.txt:442]: I fixed the wording for pull to be "of a value" and pullAll to be "of values" since that's how they differ.

File: source/fundamentals/builders/updates.txt
Comment (kyuan-mongodb) [reply source/fundamentals/builders/updates.txt:286]: I left it as it is since this is what https://mongodb.github.io/mongo-java-driver/4.2/builders/updates/ has for currentDate.  Would you like it to be changed to currentTimestamp?

Comment (jyemin) [reply source/fundamentals/builders/updates.txt:286]: Yeah, I think current docs are wrong.  currentDate helper gives you a date, currentTimestamp gives you a timestamp.

Comment (kyuan-mongodb) [reply source/fundamentals/builders/updates.txt:286]: I just tested both methods.
currentTimestamp returns: lastModified=Timestamp{value=6942526372958961671, seconds=1616432884, inc=7}
currentDate returns: lastModified=Mon Mar 22 13:08:04 EDT 2021

I think currentDate performs as expected.  So to clarify, you want currentTimestamp to replace the currentDate section?

Comment (jyemin) [reply source/fundamentals/builders/updates.txt:286]: How about two sections, one for each helper method?

**Review by jyemin — CHANGES_REQUESTED** [review]

File: source/fundamentals/builders/updates.txt
Comment (jyemin) [inline source/fundamentals/builders/updates.txt:141]: Nit: doubled-up `lastModified`?

**Review by jyemin — APPROVED** [review]
One apparent typo, but no need to re-review.

LGTM!

**Pull request merged**
